XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
+// constructed can be accessed as "this". Because init() is called during
+// construction it is not safe to blindly inspect the constructed object at the
+// moment. Though we are currently investigating how to make this safe.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
```

Comment:
The other useful bit of information is the _invocation count_.  i.e., in cases where pre()/post() can be called 0..n times, what will uniquely identify an invocation?  I agree that this can be tracked by  individually by the hooks, but it is part of the _identity_ of an invocation of a callback.  (i.e., an execution of a callback is uniquely identifed by the tuple `[asyncId, invocationId]`)